plugins/module_utils/foreman_helper.py

```diff
@@ -285,14 +287,108 @@ def ensure_resource(self, resource, entity_dict, entity, state=None, name_map=No
+            Return value:
+                Pair of boolean indicating whether something changed and the new current state if the entity
+        """
+        if state is None:
+            state = self.state
+        if entity_spec is None:
+            entity_spec = self.entity_spec
+
+        changed = False
+
+        if state == 'present_with_defaults':
+            if current_entity is None:
+                changed, current_entity = self._create_entity(resource, desired_entity, params, entity_spec)
+        elif state == 'present':
+            if current_entity is None:
+                changed, current_entity = self._create_entity(resource, desired_entity, params, entity_spec)
+            else:
+                changed, current_entity = self._update_entity(resource, desired_entity, current_entity, params, entity_spec)
+        elif state == 'absent':
+            if current_entity is not None:
+                changed, current_entity = self._delete_entity(resource, current_entity, params)
+        else:
+            self.fail_json(msg='Not a valid state: {}'.format(state))
+        return changed, current_entity
+
+    def _create_entity(self, resource, desired_entity, params, entity_spec):
+        """Create entity with given properties
+
+            Parameters:
+                resource (string): Plural name of the api resource to manipulate
+                desired_entity (dict): Desired properties of the entity
+                params (dict): Lookup parameters (i.e. parent_id for nested entities) (optional)
+                entity_spec (dict): Description of the entity structure
+            Return value:
+                Pair of boolean indicating whether something changed and the new current state if the entity
+        """
+        payload = _flatten_entity(desired_entity, entity_spec)
+        if params:
+            payload.update(params)
+        return self._resource_action(resource, 'create', payload)
+
+    def _update_entity(self, resource, desired_entity, current_entity, params, entity_spec):
+        """Update a given entity with given properties if any diverge
+
+            Parameters:
+                resource (string): Plural name of the api resource to manipulate
+                desired_entity (dict): Desired properties of the entity
+                current_entity (dict): Current properties of the entity
+                params (dict): Lookup parameters (i.e. parent_id for nested entities) (optional)
+                entity_spec (dict): Description of the entity structure
+            Return value:
+                Pair of boolean indicating whether something changed and the new current state if the entity
+        """
+        payload = {}
+        desired_entity = _flatten_entity(desired_entity, entity_spec)
+        current_entity = _flatten_entity(current_entity, entity_spec)
+        for key, value in desired_entity.items():
+            if key in current_entity and current_entity[key] == value:
```

Comment:
```suggestion
            if current_entity.get(key) == value:
```

Is that more readable? Or do we prefer the explicit `in` check? Thinking out aloud.

Reply:
What can possibly go wrong? Value will _never_ be `None` because it comes from `clean_params`.